### ISSUE TITLE
Access Denied Should include error_code and error_description

### ISSUE THREAD

**Issue by rwinch** [jzheaux/spring-security#12]
### Summary

Currently an access denied does not include an error_code or error_description

```
HTTP/1.1 403 
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Length: 0
Date: Thu, 21 Jun 2018 20:50:42 GMT
Expires: 0
Pragma: no-cache
WWW-Authenticate: Bearer
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
```

Both should be added

**Comment by jzheaux** [comment]
To be clear, this is thrown in situations like the following:

```java
@PreAuthorize("hasAuthority('SCOPE_message:read')")
public String doThis()
```

In access denied scenarios, what error code and description would you expect?

It doesn't appear straightforward how to give the user "insufficient_scope" since we'd have to inject ourselves into the expression evaluation to know if a scope was asked for in the expression. (I do have a sample of this in the resource server sandbox, using a custom SpEL bean, a la `@oauth2.hasScope('message:read')`).

So, that leaves us with invalid_request or invalid_token.

The spec says that invalid_request should respond with a 400, and invalid_token should respond with a 401: https://tools.ietf.org/html/rfc6750#section-3.1

Which is why I chose to leave it at just "Bearer" for the time being.

Here are three solutions that I've considered:

1. Just return invalid_token with a 401. I think we lose some semantic strength here since we do actually know that the token was valid, access was simply denied, and 403 communicates that nuance.
2. Just return invalid_token with a 403. The spec uses SHOULD language around the status code, not MUST.
3. Return our own error code with a 403, like "forbidden". I don't think that this is really an option, but I include it for completeness.

To get more information into the description, we are going to need to somehow inject behavior into the expression evaluation, which may mean revisiting our decision to leave this out for now. Otherwise, the description will be something like "The provided token was missing the necessary authorities to perform this request."

Thougths @rwinch @jgrandja ?

**Comment by jgrandja** [comment]
The spec states:

> insufficient_scope
         The request requires higher privileges than provided by the
         access token.  The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

Therefore, a 403 should be returned with a minimum response of `error_code: insufficient_scope`.

Seems like we'll have to hook into the ExpressionContext which will complicate things.

**Comment by jzheaux** [comment]
@jgrandja Agreed that in order to state "insufficient_scope", we need to somehow participate in expression evaluation to know that a scope is the cause of the denial.

We also need to keep in mind that not all denial scenarios boil down to insufficient scope.

In those cases (whether it is because the code doesn't know why or because the denial logic is more complex than a hasAuthority(scope)), it appears that we have two choices:

1. Depart from the spec, return a 403, and maintain the semantic strength of communicating "forbidden" as opposed to "unauthorized".
2. Stick with the spec, return a 401 - invalid_token, and lose some HTTP semantics

Aside from departing from the spec (going with the first option), this also introduces a nuance in our implementation, which a malicious user could use to learn more about the underlying system.

So, I propose the following for this first PR:

Until the code has the sophistication to know that the denial is due to insufficient scoping, we return 401 - invalid_token with a message like "The provided token lacks sufficient authority to perform this request."

We can certainly add this extra support in the future, and then start returning 403 - insufficient_scope in scenarios where we know, and stick with 401 - invalid_token when we don't.

**Comment by jzheaux** [comment]
@jgrandja Can you clarify what you mean by this:

> (the rejected scope NOT the required scope(s)).

I agree that this is a more defensive security posture; however, the spec says:

> The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

By "rejected scope" do you mean returning to the user the scopes in their JWT or do you mean the specific scope that was needed but that the JWT did not have? The spec seems to call for the latter, though it does use MAY language.

**Comment by rwinch** [comment]
I propose that we use insufficient_scope and include the scopes that exist in the description but do not use scope attribute used to protect the resource. This seems like a good compromise because nearly all of OAuth authorization is going to be based on the scopes

**Comment by jgrandja** [comment]
> Depart from the spec, return a 403, and maintain the semantic strength of communicating "forbidden" as opposed to "unauthorized".

"unauthorized" is not an option for 403 - "unauthorized" is associated to 401 when the request lacks credentials (unauthenticated) as per HTTP spec.

> Until the code has the sophistication to know that the denial is due to insufficient scoping, we return 401 - invalid_token with a message like "The provided token lacks sufficient authority to perform this request."

If the token is "authenticated" and proceeds to authorization but access is denied than we cannot return 401. It needs to be 403.


This is the way I see things. The spec defines these 3 error codes:

`invalid_request` - return 400 status in cases where a Bearer token is provided but it cannot be parsed, for example, `JwtDecoder` cannot parse the `Jwt`. This is just one example of a bad request.

`invalid_token` - return 401 status when the `Jwt` is parsed and is expired than throw `AuthenticationException` which the `BearerTokenAuthenticationEntryPoint` will pick up and respond.

`insufficient_scope` - return 403 status when the `Jwt` is parsed and not expired and proceeds to `AccessDecisionManager` and at this point access is denied for whatever reason. It could be lack of scope or it could be lack of authority based on the authorities mapping.

**Comment by jzheaux** [comment]
> nearly all of OAuth authorization is going to be based on the scopes

Agreed, we are certainly safe in optimizing for that scenario.

> I propose that we use insufficient_scope and include the scopes that exist

To clarify: "that exist on the JWT"?

If so, this is a departure from the spec as the spec says "MAY include the 'scope' attribute with the scope necessary to access the protected resource". We would be including a scope attribute that does not follow that definition. Since there is no MUST, e.g. "MAY include the 'scope' attribute which MUST contain the necessary...", it's unclear to me whether we have room to depart from that language as far as the spec is concerned. Do we?

Again, agreed that returning what scope is in the JWT is more defensive, and I don't see a security issue with it.

@rwinch It's not clear to me how to add this feature in a simple and efficient way other than to extend SecurityExpressionRoot#hasAuthority so that the scope can be inspected at that point (and we would thus know that it is due to a scope request).

Are you then proposing then that we assume insufficient_scope when access is denied, taking the (small) risk that we will incorrectly report that in more complex scenarios?

**Comment by jzheaux** [comment]
> If the token is "authenticated" and proceeds to authorization but access is denied than we cannot return 401. It needs to be 403.

@jgrandja I certainly prefer this, with the understanding that given that as a premise, we must either introduce some complexity to know why the 403 was issued or presume insufficient_scope as (I believe) @rwinch has suggested.

**Comment by jgrandja** [comment]
@jzheaux 

> the rejected scope NOT the required scope(s)

We never want to provide the caller with the list of scopes that are required to access the protected resource as this is sensitive information - they can somehow build the access token with those scopes and than try again to access the resource.

We only want to tell them that the scopes associated with the token are denied. But again, this is optional and we can just be generic and say `insufficient_scope` which I think is fine for initial implementation.

> By "rejected scope" do you mean returning to the user the scopes in their JWT

Yes

> or do you mean the specific scope that was needed but that the JWT did not have

Never

**Comment by jzheaux** [comment]
@jgrandja Totally agree that this is a more defensive posture. So then, how do you interpret this language:

> The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

It is just bad advice, or am I reading it wrong somehow?

**Comment by jgrandja** [comment]
@jzheaux 

> MAY include the "scope"
attribute with the scope necessary to access the protected
resource.

No you are reading it right. It does say **MAY**. But IMO it is also bad advice.

**Comment by rwinch** [comment]
> Are you then proposing then that we assume insufficient_scope when access is denied, taking the (small) risk that we will incorrectly report that in more complex scenarios?

Yes

> No you are reading it right. It does say MAY. But IMO it is also bad advice.

Agreed. It is poor advice from a security stance.

**Comment by jzheaux** [comment]
Sweet, @rwinch @jgrandja those are the two clarifications I needed.

I agree that we assume insufficient_scope and return the scopes on the JWT. If folks really need to get more sophisticated with their denial reporting, they can always wire a bean to use in their expressions.
